icefall/decode.py

```diff
@@ -889,3 +897,225 @@ def rescore_with_attention_decoder(
+    )
+    # nbest.fsa.scores are all 0s at this point
+
+    nbest = nbest.intersect(lattice)
+    # Now nbest.fsa has its scores set.
+    # Also, nbest.fsa inherits the attributes from `lattice`.
+    assert hasattr(nbest.fsa, "lm_scores")
+
+    am_scores = nbest.compute_am_scores()
+    ngram_lm_scores = nbest.compute_lm_scores()
+
+    # The `tokens` attribute is set inside `compile_hlg.py`
+    assert hasattr(nbest.fsa, "tokens")
+    assert isinstance(nbest.fsa.tokens, torch.Tensor)
+
+    path_to_utt_map = nbest.shape.row_ids(1).to(torch.long)
+    # the shape of memory is (T, N, C), so we use axis=1 here
+    expanded_memory = memory.index_select(1, path_to_utt_map)
+
+    if memory_key_padding_mask is not None:
+        # The shape of memory_key_padding_mask is (N, T), so we
+        # use axis=0 here.
+        expanded_memory_key_padding_mask = memory_key_padding_mask.index_select(
+            0, path_to_utt_map
+        )
+    else:
+        expanded_memory_key_padding_mask = None
+
+    # remove axis corresponding to states.
+    tokens_shape = nbest.fsa.arcs.shape().remove_axis(1)
+    tokens = k2.RaggedTensor(tokens_shape, nbest.fsa.tokens)
+    tokens = tokens.remove_values_leq(0)
+
+    device = model.device
+
+    #  import pdb
+    #
+    #  pdb.set_trace()
+    path_per_utt = (
+        nbest.shape.row_splits(1)[1:] - nbest.shape.row_splits(1)[:-1]
+    )
+    logging.info(f"path per utt: {path_per_utt}")
+    if 1 not in path_per_utt:
+        device2 = masked_lm_model.device
+
+        alignment = compute_alignment(
+            tokens.to(device2), nbest.shape.to(device2)
+        )
+        tgt_ll_list = []
+        for label_name in ["ref_labels", "hyp_labels"]:
+            (
+                masked_src_symbols,
+                src_symbols,
+                tgt_symbols,
+                src_key_padding_mask,
+                tgt_weights,
+            ) = prepare_conformer_lm_inputs(
+                alignment,
+                bos_id=sos_id,
+                eos_id=eos_id,
+                blank_id=blank_id,
+                src_label_name=label_name,
+                unmasked_weight=0.0,
+            )
+
+            masked_src_symbols = masked_src_symbols.to(torch.int64)
+            src_symbols = src_symbols.to(torch.int64)
+            tgt_symbols = tgt_symbols.to(torch.int64)
+
+            masked_lm_memory, masked_lm_pos_emb = masked_lm_model(
+                masked_src_symbols, src_key_padding_mask
+            )
+
+            tgt_nll = masked_lm_model.decoder_nll(
+                masked_lm_memory,
+                masked_lm_pos_emb,
+                src_symbols,
+                tgt_symbols,
+                src_key_padding_mask,
+            )
+
+            # nll means negative log-likelihood
+            # ll means log-likelihood
+            tgt_ll = -1 * (tgt_nll * tgt_weights).sum(dim=-1)
+
+            tgt_ll_list.append(tgt_ll)
+
+        #  tgt_ll = tgt_ll_list[1] - tgt_ll_list[0] # wer: 2.61
```

Comment:
This is a little suspicious to me; seems to me it should be hyp - ref = [1] - [0].  If the sign is wrong I think we should determine where.